File: BotsSDK/android/korebot/src/main/java/kore/botssdk/adapter/QuickRepliesTemplateAdapter.java
package kore.botssdk.adapter;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import kore.botssdk.R;
import kore.botssdk.application.AppControl;
import kore.botssdk.listener.ComposeFooterInterface;
import kore.botssdk.listener.InvokeGenericWebViewInterface;
import kore.botssdk.models.BotListModel;
import kore.botssdk.models.BotResponse;
import kore.botssdk.models.QuickReplyTemplate;
import kore.botssdk.net.SDKConfiguration;
import kore.botssdk.utils.BundleConstants;
import kore.botssdk.utils.StringUtils;
import kore.botssdk.view.viewHolder.QuickReplyViewHolder;
import kore.botssdk.view.viewUtils.RoundedCornersTransform;

public class QuickRepliesTemplateAdapter extends RecyclerView.Adapter<QuickReplyViewHolder> {

    private ArrayList<QuickReplyTemplate> quickReplyTemplateArrayList;
    Context context;
    private LayoutInflater layoutInflater;
    private RecyclerView parentRecyclerView;

    private ComposeFooterInterface composeFooterInterface;
    private InvokeGenericWebViewInterface invokeGenericWebViewInterface;
    private String quickWidgetColor,fillColor,quickReplyFontColor;
    private int dp1;
    private SharedPreferences sharedPreferences;

    public QuickRepliesTemplateAdapter(Context context, RecyclerView parentRecyclerView) {
        this.context = context;
        this.parentRecyclerView = parentRecyclerView;
        layoutInflater = LayoutInflater.from(context);
        sharedPreferences = context.getSharedPreferences(BotResponse.THEME_NAME, Context.MODE_PRIVATE);

        quickWidgetColor = SDKConfiguration.BubbleColors.quickReplyColor;
        fillColor = SDKConfiguration.BubbleColors.quickReplyColor;
        quickReplyFontColor = "#3F51B5";

        fillColor = sharedPreferences.getString(BotResponse.BUTTON_ACTIVE_BG_COLOR, fillColor);
        quickWidgetColor = sharedPreferences.getString(BotResponse.BUTTON_ACTIVE_TXT_COLOR, quickWidgetColor);
        quickReplyFontColor = sharedPreferences.getString(BotResponse.BUTTON_INACTIVE_TXT_COLOR, quickReplyFontColor);



        dp1= (int) AppControl.getInstance(context).getDimensionUtil().dp1;
    }

    @NonNull
    @Override
    public QuickReplyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View convertView = layoutInflater.inflate(R.layout.quick_replies_item_cell, null);
        GradientDrawable gradientDrawable = (GradientDrawable)convertView.findViewById(R.id.quick_reply_view).getBackground();
        gradientDrawable.setStroke(2 * dp1, Color.parseColor(quickWidgetColor));
        gradientDrawable.setColor(Color.parseColor(fillColor));
        QuickReplyViewHolder viewHolder = new QuickReplyViewHolder(convertView);
        viewHolder.getQuickReplyTitle().setTextColor(Color.parseColor(quickWidgetColor));
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull QuickReplyViewHolder holder, int position) {
        QuickReplyTemplate quickReplyTemplate = quickReplyTemplateArrayList.get(position);

        if (quickReplyTemplate.getImage_url() != null && !quickReplyTemplate.getImage_url().isEmpty()) {
            Picasso.get().load(quickReplyTemplate.getImage_url()).into(holder.getQuickReplyImage());
            holder.getQuickReplyImage().setVisibility(View.VISIBLE);
        } else {
            holder.getQuickReplyImage().setVisibility(View.GONE);
        }

        holder.getQuickReplyTitle().setText(quickReplyTemplate.getTitle());

        holder.getQuickReplyRoot().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int position =  parentRecyclerView.getChildPosition(v);
                if (composeFooterInterface != null && invokeGenericWebViewInterface != null) {
                    QuickReplyTemplate quickReplyTemplate = quickReplyTemplateArrayList.get(position);
                    if (BundleConstants.BUTTON_TYPE_POSTBACK.equalsIgnoreCase(quickReplyTemplate.getContent_type())) {
                        String quickReplyTitle = quickReplyTemplate.getTitle();
                        String quickReplyPayload = quickReplyTemplate.getPayload();
                        composeFooterInterface.onSendClick(quickReplyTitle, quickReplyPayload,false);
                    } else if(BundleConstants.BUTTON_TYPE_USER_INTENT.equalsIgnoreCase(quickReplyTemplate.getContent_type())){
                        invokeGenericWebViewInterface.invokeGenericWebView(BundleConstants.BUTTON_TYPE_USER_INTENT);
                    }else if(BundleConstants.BUTTON_TYPE_TEXT.equalsIgnoreCase(quickReplyTemplate.getContent_type())){
                        composeFooterInterface.onSendClick(quickReplyTemplate.getTitle(),quickReplyTemplate.getPayload(),false);
                    }else if(BundleConstants.BUTTON_TYPE_WEB_URL.equalsIgnoreCase(quickReplyTemplate.getContent_type())){
                        invokeGenericWebViewInterface.invokeGenericWebView(quickReplyTemplate.getPayload());
                    }else{
                        String quickReplyTitle = quickReplyTemplate.getTitle();
                        String quickReplyPayload = quickReplyTemplate.getPayload();
                        composeFooterInterface.onSendClick(quickReplyTitle, quickReplyPayload,false);
                    }
                }
            }
        });
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {

        if (quickReplyTemplateArrayList == null) {
            return 0;
        } else {
            return quickReplyTemplateArrayList.size();
        }
    }

    public void setQuickReplyTemplateArrayList(ArrayList<QuickReplyTemplate> quickReplyTemplateArrayList) {
        this.quickReplyTemplateArrayList = quickReplyTemplateArrayList;
    }

    public void setComposeFooterInterface(ComposeFooterInterface composeFooterInterface) {
        this.composeFooterInterface = composeFooterInterface;
    }

    public void setInvokeGenericWebViewInterface(InvokeGenericWebViewInterface invokeGenericWebViewInterface) {
        this.invokeGenericWebViewInterface = invokeGenericWebViewInterface;
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/adapters/QuickRepliesTemplateAdapter.kt
package com.kore.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.graphics.toColorInt
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.bitmap.CircleCrop
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.DrawableImageViewTarget
import com.kore.data.repository.preference.PreferenceRepositoryImpl
import com.kore.extensions.dpToPx
import com.kore.extensions.setRoundedCorner
import com.kore.listeners.QuickRepliesClickListener
import com.kore.model.constants.BotResponseConstants
import com.kore.model.constants.BotResponseConstants.THEME_NAME
import com.kore.ui.R

class QuickRepliesTemplateAdapter(
    private val context: Context,
    private val type: String?,
    private val quickReplies: List<Map<String, *>>,
    private val quickReliesClickListener: QuickRepliesClickListener
) : RecyclerView.Adapter<QuickRepliesTemplateAdapter.QuickReplyViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QuickReplyViewHolder {
        val convertView: View = if (type.equals(BotResponseConstants.TEMPLATE_TYPE_LIST, ignoreCase = true))
            LayoutInflater.from(context).inflate(R.layout.quick_replies_full, parent, false)
        else LayoutInflater.from(context).inflate(R.layout.quick_replies_cell, parent, false)
        return QuickReplyViewHolder(convertView)
    }

    override fun onBindViewHolder(holder: QuickReplyViewHolder, position: Int) {
        val quickReplyTemplate = quickReplies[position]
        val sharedPrefs = PreferenceRepositoryImpl()
        val bgColor = sharedPrefs.getStringValue(
            holder.quickReplyView.context, THEME_NAME, BotResponseConstants.BUBBLE_LEFT_BG_COLOR, "#ffffff"
        ).toColorInt()
        val textColor = sharedPrefs.getStringValue(
            holder.quickReplyView.context, THEME_NAME, BotResponseConstants.BUBBLE_RIGHT_BG_COLOR, "#3F51B5"
        ).toColorInt()
        holder.quickReplyView.setRoundedCorner(4.dpToPx(holder.itemView.context).toFloat())
        holder.quickReplyView.setBackgroundColor(bgColor)
        holder.quickReplyTitle.setTextColor(textColor)

        if (quickReplyTemplate[BotResponseConstants.KEY_TITLE] != null) {
            holder.quickReplyTitle.text = quickReplyTemplate[BotResponseConstants.KEY_TITLE] as String
        }
        holder.quickReplyImage.isVisible = quickReplyTemplate[BotResponseConstants.COMPONENT_TYPE_IMAGE_URL] != null
        quickReplyTemplate[BotResponseConstants.COMPONENT_TYPE_IMAGE_URL]?.let {
            Glide.with(context)
                .load(it)
                .apply(RequestOptions.bitmapTransform(CircleCrop()))
                .apply(RequestOptions().diskCacheStrategy(DiskCacheStrategy.NONE)).error(R.drawable.ic_image_photo)
                .into<DrawableImageViewTarget>(DrawableImageViewTarget(holder.quickReplyImage))
        }

        holder.quickReplyRoot.setOnClickListener {
            quickReliesClickListener.quickRepliesClicked(quickReplyTemplate)
        }
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemCount(): Int {
        return quickReplies.size
    }

    class QuickReplyViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val quickReplyView: RelativeLayout = view.findViewById(R.id.quick_reply_view)
        val quickReplyTitle: TextView = view.findViewById(R.id.quick_reply_item_text)
        val quickReplyRoot: RelativeLayout = view.findViewById(R.id.quick_reply_item_root)
        val quickReplyImage: ImageView = view.findViewById(R.id.quick_reply_item_image)
    }
}